src/Processors/QueryPlan/Optimizations/optimizeTree.cpp

```diff
@@ -117,6 +117,8 @@ void optimizeTreeSecondPass(const QueryPlanOptimizationSettings & optimization_s
     {
         optimizePrimaryKeyConditionAndLimit(stack);
 
+        tryUpdateQueryConditionCache(optimization_settings, stack);
```

Comment:
imo it'd be simpler to access cache directly from `FilterStep`, e.g. through `getGlobalContextInstance`.